### PR TITLE
vq diffusion blog post

### PR DESCRIPTION
**Pull request by williamberman**
VQ Diffusion blog post

This blog post links to relevant resources and provides examples of using the model with diffusers. It also describes the model architecture and discrete diffusion processes more broadly. It contextualizes VQ Diffusion with continuous diffusion models and autoregressive models. It then discusses further work that could be done adding more released VQ Diffusion models trained on different datasets and functionality to diffusers

### CODE REVIEW

File: vq-diffusion.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+##### AR Models
+TODO
+
+### Further steps with VQ Diffusion and 🧨 Diffusers
+
+So far, we've only ported the VQ Diffusion model trained on the ITHQ dataset. There are also released VQ Diffusion models trained on CUB-200, Oxford-102, MSCOCO, Conceptual Captions, LAION-400M, and ImageNet.
+
+VQ Diffusion also supports a faster inference strategy. The network reparameterization relies on the posterior of the diffusion process conditioned on the un-noised image being tractable. A similar formula applies when using a time stride, \\( \Delta t \\) , that skips a number of reverse diffusion steps, \\( p_\theta (x_{t - \Delta t } | x_t, y) = \sum_{\tilde{x}_0=1}^{K}{q(x_{t - \Delta t} | x_t, \tilde{x}_0)} p_\theta(\tilde{x}_0 | x_t, y) } \\).
+
+[Improved Vector Quantized Diffusion Models](https://arxiv.org/abs/2205.16007) improves upon VQ Diffusion's sample quality.
```
Comment (williamberman) [inline vq-diffusion.md:103]: TODO maybe ask for better ~two sentence descriptor about paper

File: vq-diffusion.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+- [Original Implementation](https://github.com/microsoft/VQ-Diffusion)
+- [Paper](https://arxiv.org/abs/2111.14822)
+
+### Demo
+
+🧨 Diffusers lets you run VQ Diffusion with just a few lines of code.
+
+Install dependencies
+```bash
+pip install diffusers transformers torch 'diffusers[torch]' ftfy
```
Comment (williamberman) [inline vq-diffusion.md:38]: TODO currently requires installing diffusers from git repo head

**Review by pcuenca — COMMENTED** [review]
Very cool!

File: _blog.yml [hunk 3]
```diff
@@ -1453,3 +1453,13 @@
   tags:
     - guide
     - expert-acceleration-program
+
+- local: vq-diffusion
+  title: "VQ Diffusion with 🧨 Diffusers"
+  author: williamberman
+  thumbnail: /blog/assets/117_vq_diffusion/thumbnail.png
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply _blog.yml:1460]: There's also a `_tags.yml` that you may want to use :)

File: vq-diffusion.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+![png](assets/117_vq_diffusion/vq_diffusion_teddy_bear_pool.png)
+
+### Architecture
+
+![svg](assets/117_vq_diffusion/vq_diffusion_architecture.svg)
+
+##### VQ-VAE
+
+A VQ-VAE is trained to learn the codebook of latent embedding vectors. This [blog post](https://ml.berkeley.edu/blog/posts/vq-vae/) is a good resource for understanding VQ-VAEs.
```
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:77]: absolutely!

File: vq-diffusion.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+##### Forward process
+
+In the forward diffusion process, each latent pixel can stay the same, be resampled to a different latent vector (each with equal probability), or be masked. Once a latent pixel is masked, it will stay masked. \\( \alpha_t \\), \\( \beta_t \\), and \\( \gamma_t \\) are hyperparameters that control the forward diffusion process from step \\( t-1 \\) to step \\( t \\). \\( \gamma_t \\) is the probability an unmasked pixel becomes masked. \\( \alpha_t + \beta_t \\) is the probability an unmasked pixel stays the same. The pixel can transition to any individual non-masked latent vector with a probability of \\( \beta_t \\). In other words, \\( \alpha_t + K \beta_t + \gamma_t = 1 \\) where \\( K \\) is the number of non-masked latent vectors. See section 4.1 of the paper for more details.
```
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:83]: Will replace!

Re: parameters in diagram, I can't really think of a good place for them in the diagram. Maybe as parameters for q? i.e. `q(x_t | x_{t-1} ; a, b, c)`

File: vq-diffusion.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+If you want to use FP16 weights
+```python
+from diffusers import VQDiffusionPipeline
+import torch
+
+pipe = VQDiffusionPipeline.from_pretrained("microsoft/vq-diffusion-ithq", torch_dtype=torch.float16, revision="fp16")
+```
+
+Optionally move to GPU
```
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:56]: Good point! It's definitely not fast enough to run on a cpu

File: vq-diffusion.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+##### Approximating the reverse process
+
+An encoder-decoder transformer approximates the classes of the un-noised latents, \\( x_0 \\), conditioned on the prompt, \\( y \\). The encoder is a CLIP text encoder with frozen weights. The decoder transformer provides unmasked global attention to all latent pixels and outputs the log probabilities of the categorical distribution over vector embeddings. Framing the problem as conditional sequence to sequence over discrete values provides some intuition for why the encoder-decoder transformer is a good fit. 
+
+[Taming Transformers](https://arxiv.org/abs/2012.09841) provides a good discussion on converting raw pixels to discrete tokens in a compressed latent space so that transformers become computationally feasible for image data.
```
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:89]: I unfortunately have a bad habit of using VQ-VAE and VQ-GAN interchangeably 😁. I'm relatively sure VQ-Diffusion does use the VQ-GAN from taming transformers. Will double check the paper and clarify here!

File: vq-diffusion.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+### Further steps with VQ Diffusion and 🧨 Diffusers
+
+So far, we've only ported the VQ Diffusion model trained on the ITHQ dataset. There are also released VQ Diffusion models trained on CUB-200, Oxford-102, MSCOCO, Conceptual Captions, LAION-400M, and ImageNet.
```
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:128]: I'll re-link to the README of the original repo. The weights are sometimes separate for the VQVAE vs the rest of the model so linking to their existing docs might provide the best context

File: vq-diffusion.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+prompt = "A teddy bear playing in the pool."
+
+image = pipe(prompt).images[0]
+```
+
+![png](assets/117_vq_diffusion/vq_diffusion_teddy_bear_pool.png)
+
+### Architecture
+
+![svg](assets/117_vq_diffusion/vq_diffusion_architecture.svg)
```
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:73]: So this is showing both the forward (q) and approximate reverse (p_theta) processes. I don't show the VQ-VAE encoder because we don't have to use it during inference. Masking is performed by just filling the whole latent image with the masked class (https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/vq_diffusion/pipeline_vq_diffusion.py#L253-L254 in case my explanation isn't good enought 😁 )!

LMK if those make sense and/or warrant changes to the diagram.

Separately, I'll look to make the labels bigger!

[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:73]: Updated the diagram with bigger labels!

File: vq-diffusion.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply vq-diffusion.md:89]: Oh yes you are right, lol! Sorry about that!

File: vq-diffusion.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+Despite having made tremendous strides, AR models still suffer from linear decreases in inference speed for increased image resolution, error accumulation, and directional bias. For equivalently sized AR transformer models, the big-O of VQ Diffusion's inference is better so long as the number of diffusion steps is less than the number of latent pixels. For the ITHQ dataset, the latent resolution is 32x32 and the model is trained up to 100 diffusion steps for an ~10x big-O improvement. In practice, VQ Diffusion "can be 15 times faster than AR methods while achieving a better image quality" (see [paper](https://arxiv.org/abs/2111.14822) for more details). Additionally, VQ Diffusion does not require teacher-forcing and instead learns to correct incorrectly predicted tokens. During training, noised images are both masked and have latent pixels replaced with random tokens. VQ Diffusion is also able to provide global context on \\( x_t \\) while predicting \\( x_{t-1} \\).
```
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md]: Agreed that it's highly technical and needs a higher level summary, my hope was that the sentence at the beginning of the section would serve that purpose.

We could restate the high level points as well at the end of the section but I worry it would be a bit repetitive. Thoughts?

Sentence referring to:
"While transformers have demonstrated success in AR modeling, they still suffer from linear decreases in inference speed for increased image resolution, error accumulation, and directional bias. VQ-Diffusion improves on all three pain points."

File: vq-diffusion.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+##### Approximating the reverse process
+
+An encoder-decoder transformer approximates the classes of the un-noised latents, \\( x_0 \\), conditioned on the prompt, \\( y \\). The encoder is a CLIP text encoder with frozen weights. The decoder transformer provides unmasked global attention to all latent pixels and outputs the log probabilities of the categorical distribution over vector embeddings. Framing the problem as conditional sequence to sequence over discrete values provides some intuition for why the encoder-decoder transformer is a good fit.
```
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:87]: I originally had an extra section here noting how the same head as in language modeling was being used and the dimensionality allows it to be used non-autoregressively predicting the full sequence at once. I removed it because I thought it got to in the weeds. You're definitely right that this section needs a bit of extra clarity. I'll condense that section to a sentence or two and add it back!

[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md]: Added more context here https://github.com/huggingface/blog/pull/640/commits/6fb8e2667138ec9026e352895a6c27cae3bc2994

Lmk if you think that's sufficient :)

File: _blog.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (williamberman) [reply _blog.yml:1460]: added!

File: vq-diffusion.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (williamberman) [reply vq-diffusion.md:91]: Added extra comment in the VQ-VAE noting that it uses the VQGAN variant https://github.com/huggingface/blog/pull/640/commits/ccfea450a24279db539eb4a6b3f19d700787e309

I think referring to it as a VQ-VAE in the rest of the post works a bit better since VQ-VAE is a more general term for the architecture and VQGAN is just a variant when training (iirc could be wrong here). This is how the two terms are used in the VQ-Diffusion paper

File: vq-diffusion.md (hunk 2 above)
Comment (williamberman) [reply vq-diffusion.md]: Change is committed but leaving conversation open for discussion re diffusers release!

File: vq-diffusion.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+Despite having made tremendous strides, AR models still suffer from linear decreases in inference speed for increased image resolution, error accumulation, and directional bias. For equivalently sized AR transformer models, the big-O of VQ-Diffusion's inference is better so long as the number of diffusion steps is less than the number of latent pixels. For the ITHQ dataset, the latent resolution is 32x32 and the model is trained up to 100 diffusion steps for an ~10x big-O improvement. In practice, VQ-Diffusion "can be 15 times faster than AR methods while achieving a better image quality" (see [paper](https://arxiv.org/abs/2111.14822) for more details). Additionally, VQ-Diffusion does not require teacher-forcing and instead learns to correct incorrectly predicted tokens. During training, noised images are both masked and have latent pixels replaced with random tokens. VQ-Diffusion is also able to provide global context on \\( x_t \\) while predicting \\( x_{t-1} \\).
```
Comment (patrickvonplaten) [inline vq-diffusion.md:126]: Think it's quite hard to compare diffusion models and AR image generation models in terms of complexity. 
Most AR image generation models are now a combination of GPT2 or BART-like Transformer model + VQGAN which are architecture-wise quite different from diffusion models => so hard to state that one is usually faster than the other IMO

Comment (patrickvonplaten) [reply vq-diffusion.md:126]: Also, diffusion models don't necessarily have "better image quality" IMO. E.g. https://parti.research.google/ with T5-XXL has very high image quality

File: _blog.yml [hunk 13]
```diff
@@ -1453,3 +1453,13 @@
...
+  date: November 23, 2022
```
Comment (patrickvonplaten) [inline _blog.yml:1461]:
```suggestion
  date: November 28, 2022
```

Let's merge on Monday?

**Review by patrickvonplaten — APPROVED** [review]
Super nice blog post! Really like the diagram and very nicely written. 

One small nit: In my opinion it's always difficult to say whether one architecture is always better/faster than the other. This depends on a lot of things (model architecture, hardware, size of model, ...) 

Good to merge for me!

File: vq-diffusion.md (hunk 12 above)
Comment (patrickvonplaten) [reply vq-diffusion.md:126]: Happy to leave as is though :-)

Comment (williamberman) [reply vq-diffusion.md:126]: To clarify I'm not trying to make a general diffusion to ar comparison, just vq-diffusion to ar because the architectures are both a transformer + a vqvae – with the variations discussed in the different ar cases. 

I think in that case the comparison is acceptable because the main difference during inference is the number of times the transformer is ran (the part I'm trying to emphasize with the big-O comparison). I’m sure there’s other differences in terms of transformer architecture that affect runtime but since we’re effectively looking at an order of magnitude difference in number of times the transformer has to be ran, I think it’s ok to not get too deep into that?

re: better image quality, I’m definitely not trying to make the claim that diffusion models have better image quality. The quoted excerpt from the paper is purely for VQ-Diffusion vs AR models, and (I assume) is comparing with AR models of equivalent size (for that excerpt it doesn’t explicitly say of equivalent size but I believe it’s implied and that would be consistent with the analysis in the rest of the paper).


I’m definitely open to making modifications to this section if those points aren’t clear! Any ideas on how to clarify?

Comment (williamberman) [reply vq-diffusion.md:126]: Also pls LMK if I'm missing something else in your comments but I think it sounds like we're on the same page with the general diffusion <> ar comparison and the more general quality comparison :)

File: _blog.yml (hunk 13 above)
Comment (williamberman) [reply _blog.yml:1483]: I'm not sure if we got an affirmative from Shuyang and I think we should check with @pcuenca that his feedback was all adequately addressed. Let's double check with them before merging, but besides that I'm good with whenever!

**Comment by williamberman** [comment]
> Super nice blog post! Really like the diagram and very nicely written.
> 
> One small nit: In my opinion it's always difficult to say whether one architecture is always better/faster than the other. This depends on a lot of things (model architecture, hardware, size of model, ...)
> 
> Good to merge for me (@williamberman let me know when you want to merge :-))

Good to note, Excited! 🚀

File: vq-diffusion.md (hunk 2 above)
Comment (pcuenca) [reply vq-diffusion.md]: This should work out of the box now, no need to install from `main`.

File: vq-diffusion.md [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+>
+> ... self-attention can achieve a better balance in the trade-off between the virtually unlimited receptive field of the necessarily sequential PixelRNN and the limited receptive field of the much more parallelizable PixelCNN and its various extensions.
+
+[Image Transformer](https://arxiv.org/abs/1802.05751) uses transformers by restricting self attention over local neighborhoods of pixels.
+
+[Taming Transformers](https://arxiv.org/abs/2012.09841) and [DALL-E 1](https://arxiv.org/abs/2102.12092) combine convolutions and transformers. Both train a VQ-VAE to learn a discrete latent space. A transformer is then trained with global but masked attention over the image in the compressed latent space. The transformer context is global but masked because attention is provided over all previously predicted latent pixels, but the model is still AR so attention cannot be provided over not yet predicted pixels.
```
Comment (pcuenca) [inline vq-diffusion.md:122]:
```suggestion
[Taming Transformers](https://arxiv.org/abs/2012.09841) and [DALL-E 1](https://arxiv.org/abs/2102.12092) combine convolutions and transformers. Both train a VQ-VAE to learn a discrete latent space, and then a transformer is trained in the compressed latent space. The transformer context is global but masked, because attention is provided over all previously predicted latent pixels, but the model is still AR so attention cannot be provided over not yet predicted pixels.
```

Slight rewording suggestion to avoid repetition. Please, ignore if it's not clear enough (clarity is better).

**Review by pcuenca — APPROVED** [review]
Looks good to me!

**Comment by patrickvonplaten** [comment]
Thanks a lot for adding the thumbnail @williamberman ! Let's promote today :rocket:
